CodeGeneration/Sources/generate-swift-syntax/ChildNodeChoices.swift

```diff
@@ -0,0 +1,210 @@
+        for choice in self.choices {
+          SwitchCaseSyntax("case .\(choice.enumCaseCallName)(let node):") {
+            StmtSyntax("return node.\(propertyName.nonVarCallNameOrLabelDeclName)")
+          }
+        }
+      }
+    }
+  }
+
+  func syntaxInitDecl(inputType: TypeSyntax) -> InitializerDeclSyntax {
+    try! InitializerDeclSyntax("public init?(_ node: \(inputType))") {
+      var choices = self.choices[0...]
+      switch choices.elseBody() {
+      case .codeBlock(let codeBlock):
+        codeBlock.statements
+      case .ifExpr(let ifExpr):
+        ifExpr
+      }
+    }
+  }
+}
+
+extension Node {
+  func childrenNodeChoices(forRaw: Bool = false) -> [ChildNodeChoices] {
+    if let node = self.layoutNode {
+      return node.children.compactMap { child in
+        switch child.kind {
+        case .nodeChoices(let choices):
+          return ChildNodeChoices(
+            name: child.syntaxChoicesType,
+            choices: choices.map { ChildNodeChoices.Choice($0, forRaw: forRaw) }
+          )
+        default:
+          return nil
+        }
+      }
+    } else if let node = self.collectionNode, node.elementChoices.count > 1 {
+      return [
+        ChildNodeChoices(
+          name: "Element",
+          choices: node.elementChoices.map {
+            ChildNodeChoices.Choice(SYNTAX_NODE_MAP[$0]!, forRaw: forRaw)
+          }
+        )
+      ]
+    } else {
+      return []
+    }
+  }
+}
+
+extension ChildNodeChoices.Choice {
+  var enumCaseDecl: EnumCaseDeclSyntax {
+    try! EnumCaseDeclSyntax(
+      """
+      \(self.documentation)\
+      \(self.experimentalDocNote)\
+      \(self.apiAttributes)\
+      case \(self.enumCaseDeclName)(\(self.syntaxType))
+      """
+    )
+  }
+
+  func baseTypeInitDecl(hasArgumentName: Bool) -> InitializerDeclSyntax? {
+    if let choiceType = self.choiceType, choiceType.isBase {
+      let firstName: SyntaxNodeString
+      let secondName: TokenSyntax
+      if hasArgumentName {
+        firstName = ""
+        secondName = self.labelDeclName
+      } else {
+        firstName = "_ "
+        secondName = "node"
+      }
+      return try! InitializerDeclSyntax(
+        """
+        \(self.experimentalDocNote)\
+        \(self.apiAttributes)\
+        public init(\(firstName)\(secondName): some \(choiceType.protocolType))
+        """
+      ) {
+        ExprSyntax(
+          "self = .\(self.memberCallName)(\(choiceType.syntaxType)(\(secondName)))"
+        )
+      }
+    } else {
+      return nil
+    }
+  }
+
+  var concreteTypeInitDecl: InitializerDeclSyntax {
+    try! InitializerDeclSyntax(
+      """
+      \(self.experimentalDocNote)\
+      \(self.apiAttributes)\
+      public init(_ node: \(self.syntaxType))
+      """
+    ) {
+      ExprSyntax("self = .\(self.memberCallName)(node)")
+    }
+  }
+}
+
+private extension ArraySlice where Element == ChildNodeChoices.Choice {
+  mutating func elseBody() -> IfExprSyntax.ElseBody {
```

Comment:
Mutating the array seems very surprising to me. I would prefer something that doesn’t modify the underlying array.

Reply:
I did it for performance sake, maybe I could wrap this ArraySlice api with an immutable façade under Array?

Reply:
Can’t you do something loop-based instead of recursive, like

```swift
func dunnoAboutTheName(_ choices: [ChildNodeChoices.Choice]) -> IfExprSyntax {
  precondition(!choices.isEmpty)
  var result: IfExprSyntax? = nil

  for choice in choices.reversed() {
    let elseBody: IfExprSyntax.ElseBody
    if let result {
      elseBody = .ifExpr(result)
    } else {
      elseBody = .codeBlock("return nil")
    }
    let ifExpr = IfExprSyntax(
      conditions: [
        ConditionElementSyntax(
          condition: .expression("let node = node.as(\(choice.syntaxType).self) "))
      ],
      body: CodeBlockSyntax(statements: ["self = .\(choice.memberCallName)(node)"]),
      elseKeyword: .keyword(.else),
      elseBody: elseBody
    )
  }
  return result!
}
```